zkvm quick start guide revamp (#1119)
risc0/risc0: website/api/zkvm/quickstart.md

# zkVM Quick Start

Welcome to the zkVM Quick Start page! Here are the steps to create your first proof:

1. Install the `cargo-risczero` toolchain
2. Create a new project using the `cargo-risczero` tool.
3. Familiarize yourself with the project's structure

> 4. Run your project locally

Reviewer: Can we align the outline and the list below? This overview shows 4 steps but then we walk through a 5 step process

## Concept break

The zkVM provides a zero-knowledge proof of the correct execution of Rust-based program.

> The **[host]** is responsible for launching the zkVM (or a [prover]), and then the **[guest]** is the program that runs inside the zkVM.

Reviewer: I feel like we should either bold all of the terminology here or none of it.

A proof of execution is a [receipt]; it contains a public part - a [journal] and an encryption part - a [seal].

## 1. Install the RISC Zero Toolchain

First, make sure you [install Rust], then install the RISC Zero toolchain by running the following:

```bash
cargo install cargo-binstall
cargo binstall cargo-risczero
cargo risczero install
```

## 2. Create a New Project

The `cargo-risczero` tool takes `--guest-name` parameter, a [guest] program that the zkVM runs, to generate a proof of its execution:

```bash
cargo risczero new my_project --guest-name guest_code_for_zk_proof
```

> There is a list of options in the [feature flags]. To create a proof with the zkVM on your own machine, we recommend at least 16 GB of RAM.

Reviewer 2: Since first time readers haven't seen the Bonsai section yet, maybe add something along the lines of "To avoid these hardware requirements, run proofs in [Bonsai][<link>] or skip proving with [dev-mode][<link>]"

(Sorry for the churn I know I'm suggesting further changes to my own suggestion here)

> There is a list of options in the [feature flags]. To create a proof with the zkVM on your own machine, we recommend at least 16 GB of RAM.

Reviewer: I don't see why hardware requirements belongs in the "Initialize a New Project" section, especially since both the dev-mode & bonsai workflows circumvent those hardware requirements.

> There is a list of options in the [feature flags]. To create a proof with the zkVM on your own machine, we recommend at least 16 GB of RAM.

Reviewer: also I don't think the feature flags is a very user-friendly place to land when diving into a quick start guide. 

imo the pointer to the cargo risczero docs makes more sense. (if those docs don't include all the feature flag information, we should improve the cargo docs)

Author: This quick start guide is an updated place for zkhack hackers to get all the info they need - we are unlikely going to update cargo docs today

> ## 3. Project structure

Reviewer: If we're framing this as a list of steps for users to take, "Project Structure" feels disorienting. I think step-by-step processes need to have a concrete action item. 

In terms of what that action item is -- the prior version said "edit the boiler plate." I may be out of date on the user experience, but my understanding is that users do still need to make changes to the template before they can user `cargo run`. How do new users learn what they need to do in between running `cargo risczero new` and running `cargo run --release`?

Reviewer: If it is the case that users can now run `cargo run --release` immediately after running `cargo risczero new`, then I'd suggest removing this item from quick start steps

The [Hello World tutorial] contains step-by-step instructions on how to:

- [Share private data] between the host & guest.
- [Share the guest's results] with the host publicly.
- [Generate a receipt] and read its journal contents.

## 4. Quick Development: Leveraging dev-mode

During the development of your project, running your code can take a long time due to the proof generation process. To address this issue and allow for faster iterations of your code, we suggest utilizing [dev-mode]. This mode bypasses the time-consuming proof generation process. To activate dev-mode, set the environment variable `RISC0_DEV_MODE=1` when executing your project:

```bash
RISC0_DEV_MODE=1 cargo run --release
```

## 5. Real Proof Generation

> Once you've reached a point where you're ready to generate real proofs, you can do so by setting `RISC0_DEV_MODE=0`. Generating proofs locally would be achieved by running the following:

Reviewer: This seems to be the place where users need to get warned about hardware requirements.

> Once you've reached a point where you're ready to generate real proofs, you can do so by setting `RISC0_DEV_MODE=0`. Generating proofs locally would be achieved by running the following:

Reviewer: Reading this section seems to indicate that local proof generation is the primary path and Bonsai is only worth considering if you're annoyed by slow proving times. 

I think it makes sense to revise to revise to frame Bonsai as more of a primary option. Something like:
```suggestion
Once you've reached a point where you're ready to generate real proofs, you can either proceed with local proof generation on your own machine or remote proof generation using Bonsai. 
- For local proving, ... 
- For remote proving, ... 
```

Author: we introduce bonsai in the paragraph below already

# zkVM Quick Start

Welcome to the zkVM Quick Start page! Here are the steps to create your first proof:

> 1. Install the `risczero` toolchain

Reviewer 3: ```suggestion
1. Install the `cargo-risczero` tool
```

> 2. Create a new project using the `risczero` tool.

Reviewer 3: ```suggestion
2. Create a new project using the `cargo-risczero` tool.
```

3. Familiarize yourself with the project's structure
4. Run your project locally

## Concept break

> zkVM provides a zero-knowledge proof of the correct execution of any Rust code.

Reviewer 3: ```suggestion
The zkVM provides a zero-knowledge proof of the correct execution of Rust-based programs.
```

> We want to run a piece of code and produce its proof of execution on zkVM. In the project code, we call zkVM a [prover]. The code we want to run on zkVM is a **[guest]** program. A proof of execution is a [receipt]; it contains a public part - a [journal] and an encryption part - a [seal].

Reviewer 2: "zkVM" almost always takes the article "the" (or occasionally "a" to refer to an instance of the zkVM). It's never used without an article.
```suggestion
We want to run a piece of code and produce its proof of execution on the zkVM. In the project code, we call the zkVM a [prover]. The code we want to run on zkVM is a **[guest]** program. A proof of execution is a [receipt]; it contains a public part - a [journal] and an encryption part - a [seal].
```

> We want to run a piece of code and produce its proof of execution on zkVM. In the project code, we call zkVM a [prover]. The code we want to run on zkVM is a **[guest]** program. A proof of execution is a [receipt]; it contains a public part - a [journal] and an encryption part - a [seal].

Reviewer 2: Personally, I find it a bit odd to "call the zkVM a prover" -- hence the different names and different definitions in the terminology section -- but perhaps this is a helpful simplification here.

> zkVM runs on the **[host]**, which loads the guest program to the zkVM and provides input to the guest. zkVM, or a prover, proves the guest execution by producing a locally verified receipt.

Reviewer 3: This section seems a bit confusing to me. I think we should make it clear that there's the zkVM, in which the host is responsible for launching the zkVM, and then the guest is the program that runs inside of the zkVM.

zkVM provides a zero-knowledge proof of the correct execution of any Rust code.
We want to run a piece of code and produce its proof of execution on zkVM. In the project code, we call zkVM a [prover]. The code we want to run on zkVM is a **[guest]** program. A proof of execution is a [receipt]; it contains a public part - a [journal] and an encryption part - a [seal].

zkVM runs on the **[host]**, which loads the guest program to the zkVM and provides input to the guest. zkVM, or a prover, proves the guest execution by producing a locally verified receipt.

## 1. Install the RISC Zero Toolchain

If you have already installed Rust via `rustup`, you can install our toolchain by running
First, make sure you [install Rust], then install the RISC Zero toolchain by running the following:

```bash
cargo install cargo-binstall
cargo binstall cargo-risczero
cargo risczero install
```

> If you need to install Rust or encounter problems, take a look at our [full installation instructions](install.md).

Reviewer 2: It feels a bit odd to me to have a zkVM installation page and not refer to it in our zkVM quickstart. And that page does have more information than this one. If you want to drop this line can you reference those instructions in some other way?

> The `cargo-risczero` tool takes `--guest-name` parameter, a [guest] program that the zkVM runs, to generate a proof of its execution:

Reviewer 2: Maybe it's just me, but I'm worried this line will confuse people that the `--guest-name` parameter specifies an existing guest program. They will still need to write their guest code, this just gives it a name.

> To run the zkVM on your own machine, we recommend at least 16 GB of RAM.

Reviewer 3: I think it's ambigious what "run the zkVM" means. It could mean just the executor (aka dev-mode) in which case not much RAM is required. It could mean generating a proof, in which case having lots of memory will be helpful.

Reviewer 2: Perhaps
```suggestion
To create a proof with the zkVM on your own machine, we recommend at least 16 GB of RAM.
```

## 3. Project structure

The [Hello World tutorial] contains step-by-step instructions on how to:

- [Share private data] between the host & guest.
- [Share the guest's results] with the host publicly.
- [Generate a receipt] and read its journal contents.

## 4. Quick Development: Leveraging Dev Mode

> During the development of your project, running your code can take a long time due to the proof generation process. To address this issue and allow for faster iterations of your code, we suggest utilizing [Dev Mode]. This mode bypasses the time-consuming proof generation process. To activate dev-mode, set the environment variable `RISC0_DEV_MODE=true` when executing your project:

Reviewer 3: ```suggestion
During the development of your project, running your code can take a long time due to the proof generation process. To address this issue and allow for faster iterations of your code, we suggest utilizing [Dev Mode]. This mode bypasses the time-consuming proof generation process. To activate dev-mode, set the environment variable `RISC0_DEV_MODE=1` when executing your project:
```

```bash
RISC0_DEV_MODE=true cargo run --release
```

## 5. Real Proof Generation

> Once you've reached a point where you're ready to generate real proofs, you can do so by setting the `RISC0_DEV_MODE=false`. We recommend that you specify the feature flag `disable-dev-mode`, which will ensure that dev-mode is not accidentally enabled. Generating proofs locally would be achieved by running the following:

Reviewer 3: ```suggestion
Once you've reached a point where you're ready to generate real proofs, you can do so by setting `RISC0_DEV_MODE=false`. We recommend that you specify the feature flag `disable-dev-mode`, which will ensure that dev-mode is not accidentally enabled. Generating proofs locally would be achieved by running the following:
```

Reviewer 3: I don't now that I'd recommend using `disable-dev-mode`. But perhaps we could mention that it's a feature available if someone needs it.

```bash
RISC0_DEV_MODE=false cargo run --release --features disable-dev-mode
```

> Note that since proofs are now being generated, the execution time of your project will be significantly longer than when running in dev mode. Consider using [Bonsai] to generate proofs remotely, as it will likely be significantly faster than running proofs locally. You can [request access] to Bonsai to set additional flags.

Reviewer 3: ```suggestion
Note that since proofs are now being generated, the execution time will be significantly longer than when running in dev-mode. Consider using [Bonsai] to generate proofs remotely, as it will likely be significantly faster than running proofs locally. You can [request access] to Bonsai to set additional flags.
```

### Executor Statistics

To gain insights into your application's performance, you can obtain executor statistics by setting the `RUST_LOG` environment variable to `"executor=info"`.

> Setting this filter will print statistics about the execution before proof generation, so you can understand how computationally expensive your application is. Since the statistics concern only the executor phase, it is recommended to run your application in dev mode to avoid the overhead of proof generation:

Reviewer 3: ```suggestion
Setting this filter will print statistics about the execution before proof generation, so you can understand how computationally expensive your application is. Since the statistics concern only the executor phase, it is recommended to run your application in dev-mode to avoid the overhead of proof generation:
```

```bash

> RISC0_DEV_MODE=true RUST_LOG="executor=info" cargo run --release

Reviewer 3: ```suggestion
RISC0_DEV_MODE=1 RUST_LOG="executor=info" cargo run --release
```

- Segments Count
- Execution time

> Knowing these statistics is helpful for estimating the cost of your application before submitting real workloads to Bonsai, as the cost of proof generation is proportional to the number of cycles and segments used.

Reviewer 3: ```suggestion
Knowing these statistics is useful for estimating the cost of your application before submitting real workloads to Bonsai, as the cost of proof generation is proportional to the number of cycles and segments used.
```

### Other options

> There is a list of options in the [feature flags].

Reviewer 3: I think we should drop this whole section if it's just a single sentence.

Knowing these statistics is useful for estimating the cost of your application before submitting real workloads to Bonsai, as the cost of proof generation is proportional to the number of cycles and segments used.

> **Congratulations!**
>
> _That's all it takes to build and run a minimal RISC Zero application._

[Bonsai]: ../bonsai/quickstart

> [dev-mode]: dev-mode.md

Reviewer 3: ```suggestion
[dev-mode]: ./dev-mode.md
```
